### ISSUE TITLE
[FSDP] `summon_full_params(offload_to_cpu=True)` keeps `_full_param_padded` on GPU

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#83076]
TL;DR: `summon_full_params(offload_to_cpu=True)` does not free the GPU copy of the unsharded flattened parameter, so both a CPU and GPU copy exist when in the context.

---
Repro:
```
def test_cpu_offload_full_param_padded(self):
    model = nn.Linear(5, 5).cuda()
    fsdp_model = FSDP(model)
    with FSDP.summon_full_params(fsdp_model, offload_to_cpu=True):
        ...
```
(1) Add a `print()` like
```
print(f"[Rank {self.rank}] _full_param_padded: {p._full_param_padded.device} {p._full_param_padded.storage().size()}")
```
after this line:
https://github.com/pytorch/pytorch/blob/6c60a656b02fbd09661b282bec53940b184db3ca/torch/distributed/fsdp/fully_sharded_data_parallel.py#L2563

(2) Add a `print()` like
```
print(f"[Rank {self.rank}] _full_param_padded: {p._full_param_padded.device} {p._full_param_padded.storage().size()}")
```
before this line:
https://github.com/pytorch/pytorch/blob/6c60a656b02fbd09661b282bec53940b184db3ca/torch/distributed/fsdp/fully_sharded_data_parallel.py#L2594

We see a print like:
```
[Rank 0] _full_param_padded: cuda:0 30
[Rank 0] _full_param_padded: cuda:0 30
[Rank 1] _full_param_padded: cuda:1 30
[Rank 1] _full_param_padded: cuda:1 30
```
---

I think that the move to CPU should be done in-place to allow the GPU copy of the (padded) unsharded flattened parameter to be freed. Otherwise, full model state dict may unexpectedly hit an OOM with `offload_to_cpu=True`. (Full model state dict will recurse down the module tree, so at the leaf module, all parent modules' parameters are unsharded, representing the peak memory usage. For large models, this cannot fit in GPU, so `offload_to_cpu=True` is necessary.)

Like the other `summon_full_params()` issue, I am addressing this in my refactoring, so if this is not blocking, there is no need to land an intermediate fix.

cc @zhaojuanmao @mrshenli @rohan-varma @ezyang

**Comment by awgu** [comment]
More detailed repro:
```
def test_cpu_offload_full_param_padded(self):
    model = nn.Sequential(
        nn.Linear(10, 10, bias=False),
        nn.Sequential(
            nn.Linear(10, 10, bias=False),
            nn.Sequential(
                nn.Linear(10, 10, bias=False),
                nn.Sequential(
                    nn.Linear(10, 10, bias=False),
                )
            )
        ),
    ).cuda()
    print(f"[Rank {self.rank}] torch.cuda.memory_allocated: {torch.cuda.memory_allocated(self.rank)}")
    torch.cuda.empty_cache()
    import functools
    auto_wrap_policy = functools.partial(
        transformer_auto_wrap_policy,
        transformer_layer_cls={nn.Sequential},
    )
    fsdp_model = FSDP(model, auto_wrap_policy=auto_wrap_policy)
    if self.rank == 0:
        print(fsdp_model)
    print(f"[Rank {self.rank}] torch.cuda.memory_allocated: {torch.cuda.memory_allocated(self.rank)}")
    torch.cuda.empty_cache()
    with FSDP.summon_full_params(fsdp_model, offload_to_cpu=True):
        print(f"[Rank {self.rank}] torch.cuda.memory_allocated: {torch.cuda.memory_allocated(self.rank)}")
```
Trunk:
```
[Rank 0] torch.cuda.memory_allocated: 2048
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): Sequential(
      (0): Linear(in_features=10, out_features=10, bias=False)
      (1): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Sequential(
            (0): Linear(in_features=10, out_features=10, bias=False)
            (1): FullyShardedDataParallel(
              (_fsdp_wrapped_module): FlattenParamsWrapper(
                (_fpw_module): Sequential(
                  (0): Linear(in_features=10, out_features=10, bias=False)
                  (1): FullyShardedDataParallel(
                    (_fsdp_wrapped_module): FlattenParamsWrapper(
                      (_fpw_module): Sequential(
                        (0): Linear(in_features=10, out_features=10, bias=False)
                      )
                    )
                  )
                )
              )
            )
          )
        )
      )
    )
  )
)
[Rank 0] torch.cuda.memory_allocated: 2048
[Rank 0] torch.cuda.memory_allocated: 4096
[Rank 1] torch.cuda.memory_allocated: 2048
[Rank 1] torch.cuda.memory_allocated: 2048
[Rank 1] torch.cuda.memory_allocated: 4096
```
Refactored:
```
[Rank 1] torch.cuda.memory_allocated: 2048
[Rank 1] torch.cuda.memory_allocated: 2048
[Rank 1] torch.cuda.memory_allocated: 2048
[Rank 0] torch.cuda.memory_allocated: 2048
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): Sequential(
      (0): Linear(in_features=10, out_features=10, bias=False)
      (1): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Sequential(
            (0): Linear(in_features=10, out_features=10, bias=False)
            (1): FullyShardedDataParallel(
              (_fsdp_wrapped_module): FlattenParamsWrapper(
                (_fpw_module): Sequential(
                  (0): Linear(in_features=10, out_features=10, bias=False)
                  (1): FullyShardedDataParallel(
                    (_fsdp_wrapped_module): FlattenParamsWrapper(
                      (_fpw_module): Sequential(
                        (0): Linear(in_features=10, out_features=10, bias=False)
                      )
                    )
                  )
                )
              )
            )
          )
        )
      )
    )
  )
)
[Rank 0] torch.cuda.memory_allocated: 2048
[Rank 0] torch.cuda.memory_allocated: 2048

```

**Comment by awgu** [comment]
https://github.com/pytorch/pytorch/pull/83665
